File: sosy-backend/app/schemas/matching.py
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field

# User Profile Schemas
class UserProfileBase(BaseModel):
    wp_user_id: int
    social_energy: Optional[str] = None
    conversation_style: Optional[str] = None
    social_goal: Optional[str] = None
    group_size_preference: Optional[int] = None
    gender: Optional[str] = None
    gender_preference: Optional[str] = None
    activity_types: Optional[List[str]] = None
    discussion_topics: Optional[List[str]] = None
    life_stage: Optional[str] = None
    cultural_background: Optional[str] = None
    price_tier: Optional[str] = None

class UserProfileCreate(UserProfileBase):
    pass

class UserProfileUpdate(BaseModel):
    social_energy: Optional[str] = None
    conversation_style: Optional[str] = None
    social_goal: Optional[str] = None
    group_size_preference: Optional[int] = None
    gender: Optional[str] = None
    gender_preference: Optional[str] = None
    activity_types: Optional[List[str]] = None
    discussion_topics: Optional[List[str]] = None
    life_stage: Optional[str] = None
    cultural_background: Optional[str] = None
    price_tier: Optional[str] = None
    reliability_score: Optional[float] = None
    attendance_rate: Optional[float] = None

class UserProfile(UserProfileBase):
    id: int
    reliability_score: float
    attendance_rate: float
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

# Matching Session Schemas
class MatchingSessionCreate(BaseModel):
    event_id: int
    target_group_size: int = Field(..., description="4 for deep, 6 for casual")
    conversation_style: str = Field(..., description="deep or casual")

class MatchingSession(BaseModel):
    id: int
    event_id: int
    event_name: Optional[str]
    session_date: Optional[datetime]
    status: str
    target_group_size: int
    conversation_style: str
    created_at: datetime

    class Config:
        from_attributes = True

# Matching Group Schemas
class GroupMember(BaseModel):
    user_id: int
    username: str
    email: str
    display_name: Optional[str]
    social_energy: Optional[str]
    match_score: Optional[float]
    has_personality_test: bool

class MatchingGroup(BaseModel):
    id: int
    session_id: int
    group_number: int
    group_size: int
    average_match_score: float
    members: List[GroupMember]
    created_at: datetime

    class Config:
        from_attributes = True

# Match Score Detail
class MatchScoreDetail(BaseModel):
    user1_id: int
    user1_name: str
    user2_id: int
    user2_name: str
    social_energy_score: float
    conversation_style_score: float
    social_goal_score: float
    group_size_score: float
    gender_comfort_score: float
    interest_score: float
    life_context_score: float
    cultural_score: float
    financial_score: float
    reliability_score: float
    total_match_score: float
    matching_criteria_count: int

# Matching Result
class MatchingResult(BaseModel):
    session: MatchingSession
    groups: List[MatchingGroup]
    total_users: int
    matched_users: int
    unmatched_users: int
    average_group_score: float

# Energy Feedback
class EnergyFeedbackCreate(BaseModel):
    group_id: int
    user_id: int
    rated_user_id: int
    energy_impact: str = Field(..., description="energized, neutral, or drained")
    rating: int = Field(..., ge=1, le=5)
    feedback_text: Optional[str] = None

class EnergyFeedback(BaseModel):
    id: int
    group_id: int
    user_id: int
    rated_user_id: int
    energy_impact: str
    rating: int
    feedback_text: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True

File: sosy-frontend/src/lib/api/matching.ts
import { apiClient } from '../api';

export interface UserProfile {
  id: number;
  wp_user_id: number;
  social_energy: string | null;
  conversation_style: string | null;
  social_goal: string | null;
  group_size_preference: number | null;
  gender: string | null;
  gender_preference: string | null;
  activity_types: string[] | null;
  discussion_topics: string[] | null;
  life_stage: string | null;
  cultural_background: string | null;
  price_tier: string | null;
  reliability_score: number;
  attendance_rate: number;
  created_at: string;
  updated_at: string | null;
}

export interface UserProfileCreate {
  wp_user_id: number;
  social_energy?: string;
  conversation_style?: string;
  social_goal?: string;
  group_size_preference?: number;
  gender?: string;
  gender_preference?: string;
  activity_types?: string[];
  discussion_topics?: string[];
  life_stage?: string;
  cultural_background?: string;
  price_tier?: string;
}

export interface UserProfileUpdate {
  social_energy?: string;
  conversation_style?: string;
  social_goal?: string;
  group_size_preference?: number;
  gender?: string;
  gender_preference?: string;
  activity_types?: string[];
  discussion_topics?: string[];
  life_stage?: string;
  cultural_background?: string;
  price_tier?: string;
  reliability_score?: number;
  attendance_rate?: number;
}

export interface MatchingSession {
  id: number;
  event_id: number;
  event_name: string | null;
  session_date: string | null;
  status: string;
  target_group_size: number;
  conversation_style: string;
  created_at: string;
}

export interface GroupMember {
  user_id: number;
  username: string;
  email: string;
  display_name: string | null;
  social_energy: string | null;
  match_score: number | null;
  has_personality_test: boolean;
}

export interface MatchingGroup {
  id: number;
  session_id: number;
  group_number: number;
  group_size: number;
  average_match_score: number;
  members: GroupMember[];
  created_at: string;
}

export interface MatchingResult {
  session: MatchingSession;
  groups: MatchingGroup[];
  total_users: number;
  matched_users: number;
  unmatched_users: number;
  average_group_score: number;
}

export interface MatchScoreDetail {
  user1_id: number;
  user1_name: string;
  user2_id: number;
  user2_name: string;
  social_energy_score: number;
  conversation_style_score: number;
  social_goal_score: number;
  group_size_score: number;
  gender_comfort_score: number;
  interest_score: number;
  life_context_score: number;
  cultural_score: number;
  financial_score: number;
  reliability_score: number;
  total_match_score: number;
  matching_criteria_count: number;
}

export interface EnergyFeedbackCreate {
  group_id: number;
  user_id: number;
  rated_user_id: number;
  energy_impact: 'energized' | 'neutral' | 'drained';
  rating: number;
  feedback_text?: string;
}

export interface EnergyFeedback {
  id: number;
  group_id: number;
  user_id: number;
  rated_user_id: number;
  energy_impact: string;
  rating: number;
  feedback_text: string | null;
  created_at: string;
}

// ==================== API Functions ====================

export const matchingApi = {
  // User Profile APIs
  async getUserProfile(wpUserId: number): Promise<UserProfile> {
    const response = await apiClient.get(`/matching/profiles/${wpUserId}`);
    return response.data;
  },

  async createUserProfile(data: UserProfileCreate): Promise<UserProfile> {
    const response = await apiClient.post('/matching/profiles', data);
    return response.data;
  },

  async updateUserProfile(
    wpUserId: number,
    data: UserProfileUpdate
  ): Promise<UserProfile> {
    const response = await apiClient.put(`/matching/profiles/${wpUserId}`, data);
    return response.data;
  },

  // Matching APIs
  async createMatching(
    eventId: number,
    targetGroupSize: number = 4,
    conversationStyle: 'deep' | 'casual' = 'deep'
  ): Promise<MatchingResult> {
    const response = await apiClient.post(
      `/matching/events/${eventId}/match`,
      null,
      {
        params: {
          target_group_size: targetGroupSize,
          conversation_style: conversationStyle,
        },
      }
    );
    return response.data;
  },

  async getMatchingSession(sessionId: number): Promise<MatchingResult> {
    const response = await apiClient.get(`/matching/sessions/${sessionId}`);
    return response.data;
  },

  async getEventSessions(eventId: number): Promise<MatchingSession[]> {
    const response = await apiClient.get(`/matching/events/${eventId}/sessions`);
    return response.data;
  },

  async getGroupMatchScores(groupId: number): Promise<MatchScoreDetail[]> {
    const response = await apiClient.get(`/matching/groups/${groupId}/scores`);
    return response.data;
  },

  // Energy Feedback APIs
  async createEnergyFeedback(data: EnergyFeedbackCreate): Promise<EnergyFeedback> {
    const response = await apiClient.post('/matching/feedback', data);
    return response.data;
  },
};